### PR TITLE
Adding Spatial Mouse to MRTK3

### PR DESCRIPTION
**Pull request by marlenaklein-msft**
## Overview
Adds a new controller that can be added to the MRTK XR Rig to enable a spatial mouse. This is an experimental feature.

Updates #10690. Main changes: fix Z-scroll bug, fix last known depth bug, modify interactor to work with custom cursors, and move the spatial mouse input action map bindings into the default MRTK input action map, 

## Changes
- Fixes: #10688.

### CODE REVIEW

File: com.microsoft.mrtk.input/Experimental/SpatialMouse/Interactor/SpatialMouseInteractor.cs [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+using UnityEngine.XR.Interaction.Toolkit;
+using UnityEngine.XR.Interaction.Toolkit.Inputs;
+using UnityEngine.XR.Interaction.Toolkit.UI;
+
+namespace Microsoft.MixedReality.Toolkit.Input.Experimental
+{
+    /// <summary>
+    /// An extension of XRRayInteractor which has extra functionality for handling spatial mouse movement and scrolling. 
+    /// </summary>
+    /// <remarks>
```
Comment (AMollis) [inline com.microsoft.mrtk.input/Experimental/SpatialMouse/Interactor/SpatialMouseInteractor.cs:18]: Remove extra "<remarks>" tag

File: com.microsoft.mrtk.input/Experimental/SpatialMouse/Interactor/SpatialMouseInteractor.cs [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        [SerializeField]
+        [Tooltip("The scale factor to apply to the mouse deltas.")]
+        public float mouseSensitivity = .05f;
+
+        /// <summary>
+        /// The scale factor to apply to the mouse wheel.
+        /// </summary>
+        [SerializeField]
+        [Tooltip("The scale factor to apply to the mouse wheel.")]
+        public float mouseWheelSensitivity = .002f;
```
Comment (AMollis) [inline com.microsoft.mrtk.input/Experimental/SpatialMouse/Interactor/SpatialMouseInteractor.cs:50]: make " public float mouseWheelSensitivity "  private.

File: com.microsoft.mrtk.input/Experimental/SpatialMouse/Interactor/SpatialMouseInteractor.cs [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        [SerializeField]
+        [Tooltip("The Input System action used for mouse scrolling. Must be a Vector2 Control.")]
+        private InputActionProperty mouseScrollAction;
+
+        /// <summary>
+        /// The scale factor to apply to the mouse deltas.
+        /// </summary>
+        [SerializeField]
+        [Tooltip("The scale factor to apply to the mouse deltas.")]
+        public float mouseSensitivity = .05f;
```
Comment (AMollis) [inline com.microsoft.mrtk.input/Experimental/SpatialMouse/Interactor/SpatialMouseInteractor.cs:43]: make "public float mouseSensitivity" private

File: com.microsoft.mrtk.input/Experimental/SpatialMouse/Interactor/SpatialMouseInteractor.cs [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    /// to have a lot of value even if the details haven�t fully been fleshed out. 
+    /// For these types of features, we want the community to see them and get 
+    /// value out of them early enough so to provide feedback. 
+    /// </remarks>
+    [AddComponentMenu("Scripts/Microsoft/MRTK/Input/MRTK Spatial Mouse Interactor")]
+    public class SpatialMouseInteractor : XRRayInteractor, IRayInteractor
+    {
+        [SerializeField, Experimental]
+        [Tooltip("The Input System action used for mouse movement. Must be a Vector2 Control.")]
+        private InputActionProperty mouseMoveAction;
```
Comment (AMollis) [inline com.microsoft.mrtk.input/Experimental/SpatialMouse/Interactor/SpatialMouseInteractor.cs:32]: Please follow the pattern used elsewhere in MRTK3.   Create "private serializable fields" and "public properties" to expose these fields.

For example

```
        [SerializeField, Experimental]
        [Tooltip("The Input System action used for mouse movement. Must be a Vector2 Control.")]
        private InputActionProperty mouseMoveAction;

        /// <summary>
        /// The Input System action used for mouse movement. Must be a Vector2 Control.
        /// </summary>
        public InputActionProperty MouseMoveAction 
        {
             get => mouseMoveAction;
             set => mouseMoveAction = value;
        }
```

Alternatively, you can create serializable public property with the "field" attribute.

```
        /// <summary>
        /// The Input System action used for mouse movement. Must be a Vector2 Control.
        /// </summary>
        [field: SerializeField, Tooltip(" The Input System action used for mouse movement. Must be a Vector2 Control..")]
         public InputActionProperty MouseMoveAction  { get; set; }
```

**Review by AMollis — CHANGES_REQUESTED** [review]
Still reviewing...but here's some comments

File: com.microsoft.mrtk.input/Experimental/SpatialMouse/Interactor/SpatialMouseInteractor.cs [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+        public float mouseHideThreshold = 20.0f;
+
+        /// <summary>
+        /// he time (in seconds) of no mouse activity before reseting the mouse cursor to the center of the FoV.
+        /// </summary>
+        [SerializeField]
+        [Tooltip("The time (in seconds) of no mouse activity before reseting the mouse cursor to the center of the FoV.")]
+        public float mouseResetThreshold = 0.2f;
+
+        private CursorLockMode restoreLockState;
```
Comment (AMollis) [inline com.microsoft.mrtk.input/Experimental/SpatialMouse/Interactor/SpatialMouseInteractor.cs:66]: can you please group all the private non-serialized fields to the top of the class, to improve readability.

File: com.microsoft.mrtk.input/Experimental/SpatialMouse/Interactor/SpatialMouseInteractor.cs [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+            if (mouseScrollAction != null)
+            {
+                mouseScrollAction.action.performed += OnMouseScroll;
+                mouseScrollAction.EnableDirectAction();
+            }
+
+            restoreLockState = Cursor.lockState;
+            Cursor.lockState =  CursorLockMode.Locked;
+        }
+
```
Comment (AMollis) [inline com.microsoft.mrtk.input/Experimental/SpatialMouse/Interactor/SpatialMouseInteractor.cs:93]: add `/// <inheritdoc />`  so API reference documents pull in base class summaries

File: com.microsoft.mrtk.input/Experimental/SpatialMouse/Interactor/SpatialMouseInteractor.cs [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+        private CursorLockMode restoreLockState;
+
+        private float timeSinceLastMouseEvent = Mathf.Infinity;
+
+        /// <summary>
+        /// Returns true if mouse is actively in use. Should be used to show/hide mouse-specific feedback (e.g. cursor).
+        /// </summary>
+        public bool IsInUse => hasSelection || (timeSinceLastMouseEvent < mouseHideThreshold);
+
+        protected override void OnEnable()
```
Comment (AMollis) [inline com.microsoft.mrtk.input/Experimental/SpatialMouse/Interactor/SpatialMouseInteractor.cs:75]: add `/// <inheritdoc />`  so API reference documents pull in base class summaries

File: com.microsoft.mrtk.input/Experimental/SpatialMouse/Interactor/SpatialMouseInteractorCursorVisual.cs [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+    /// subject to change over time. Parts of the MRTK, such as this class, appear 
+    /// to have a lot of value even if the details haven�t fully been fleshed out. 
+    /// For these types of features, we want the community to see them and get 
+    /// value out of them early enough so to provide feedback. 
+    /// </remarks>
+    public class SpatialMouseInteractorCursorVisual : BaseReticleVisual
+    {
+        [SerializeField, Experimental]
+        [Tooltip("The ray interactor which this visual represents.")]
+        private SpatialMouseInteractor mouseInteractor;
```
Comment (AMollis) [inline com.microsoft.mrtk.input/Experimental/SpatialMouse/Interactor/SpatialMouseInteractorCursorVisual.cs:27]: Please follow the pattern used elsewhere in MRTK3.   Create "private serializable fields" and "public properties" to expose these fields.

For example

```
        [SerializeField, Experimental]
        [Tooltip("The Input System action used for mouse movement. Must be a Vector2 Control.")]
        private InputActionProperty mouseMoveAction;

        /// <summary>
        /// The Input System action used for mouse movement. Must be a Vector2 Control.
        /// </summary>
        public InputActionProperty MouseMoveAction 
        {
             get => mouseMoveAction;
             set => mouseMoveAction = value;
        }
```

Alternatively, you can create serializable public property with the "field" attribute.

```
        /// <summary>
        /// The Input System action used for mouse movement. Must be a Vector2 Control.
        /// </summary>
        [field: SerializeField, Tooltip(" The Input System action used for mouse movement. Must be a Vector2 Control..")]
         public InputActionProperty MouseMoveAction  { get; set; }
```

File: com.microsoft.mrtk.input/Experimental/SpatialMouse/Interactor/SpatialMouseInteractorCursorVisual.cs [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+
+
+        [SerializeField]
+        [Tooltip("The default distance of the reticle (cursor)")]
+        private float defaultDistance = 1.0f;
+        
+        /// <summary>
+        /// See <see cref="MonoBehaviour"/>.
+        /// </summary>
+        protected void OnEnable()
```
Comment (AMollis) [inline com.microsoft.mrtk.input/Experimental/SpatialMouse/Interactor/SpatialMouseInteractorCursorVisual.cs:37]: make `OnEnabled()` private.

File: com.microsoft.mrtk.input/Experimental/SpatialMouse/Interactor/SpatialMouseInteractorCursorVisual.cs [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+        {
+            mouseInteractor.selectEntered.AddListener(LocateTargetHitPoint);
+
+            Application.onBeforeRender += OnBeforeRenderCursor;
+        }
+
+        /// <summary>
+        /// See <see cref="MonoBehaviour"/>.
+        /// </summary>
+        protected void OnDisable()
```
Comment (AMollis) [inline com.microsoft.mrtk.input/Experimental/SpatialMouse/Interactor/SpatialMouseInteractorCursorVisual.cs:47]: make `OnDisable()` private.

**Review by AMollis — CHANGES_REQUESTED** [review]
Some more changes requeseted

File: com.microsoft.mrtk.input/Experimental/SpatialMouse/Interactor/SpatialMouseInteractorCursorVisual.cs (hunk 10 above)
Comment (keveleigh) [reply com.microsoft.mrtk.input/Experimental/SpatialMouse/Interactor/SpatialMouseInteractorCursorVisual.cs:47]: We'd been moving to using `protected` over `private` for Unity-powered methods so they're not accidentally hidden for derived classes. I'd recommend not changing this.

https://github.com/microsoft/Microsoft.Unity.Analyzers/blob/main/doc/UNT0021.md

Comment (AMollis) [reply com.microsoft.mrtk.input/Experimental/SpatialMouse/Interactor/SpatialMouseInteractorCursorVisual.cs:47]: Thanks Kurtis for pointing to this information.    Given https://github.com/microsoft/Microsoft.Unity.Analyzers/blob/main/doc/UNT0021.md ,  I aggree should be made protect.

But @keveleigh ... how does this help a derived class not hide the base behaviour?  The base class would need to mark the function as `protect virtual` for the derived class to not hide the base behaviour...correct?

Comment (keveleigh) [reply com.microsoft.mrtk.input/Experimental/SpatialMouse/Interactor/SpatialMouseInteractorCursorVisual.cs]: You're right that it doesn't prevent the derived class from hiding the base implementation, but it at least allows the compiler to complain that you _are_ hiding the base implementation. If they're both `private`, the compiler is fine with it, but (silently) only the derived class' method gets called. You're correct on needing to mark it `virtual` if we want to allow overriding, though I don't know if we necessarily do in _all_ cases.

For example:
image

image

Comment (AMollis) [reply com.microsoft.mrtk.input/Experimental/SpatialMouse/Interactor/SpatialMouseInteractorCursorVisual.cs]: Got it....making this `protected` is about exposing the error/warning message.

I think I would rather do `protected virtual` to allow for extensions.  Just because it "feels" better than just making it `protected`.   Having a protected non-virtual method, in non-Unity circumstances, invites derived classes to call this function and not extended or override it.  Since an app shouldn't be calling these "Unity event functions", making it `protected virtual` at least expressions a different intent...you can extend or override the function's behavior.

**Review by AMollis — APPROVED** [review]

File: com.microsoft.mrtk.input/Experimental/SpatialMouse/MRTK Gaze Controller.prefab
Comment (keveleigh) [inline com.microsoft.mrtk.input/Experimental/SpatialMouse/MRTK Gaze Controller.prefab:1]: Why was this moved? Can it please be moved back?

Comment (marlenaklein-msft) [reply com.microsoft.mrtk.input/Experimental/SpatialMouse/MRTK Gaze Controller.prefab:1]: ah! yes this was an accident, will fix

Comment (keveleigh) [reply com.microsoft.mrtk.input/Experimental/SpatialMouse/MRTK Gaze Controller.prefab:1]: Thank you!

Comment (keveleigh) [reply com.microsoft.mrtk.input/Experimental/SpatialMouse/MRTK Gaze Controller.prefab:1]: No worries! Thank you!
